src/cast_core/src/CrudActor.cpp

```diff
@@ -1206,21 +1296,125 @@ struct CrudActor::PhaseConfig {
+            std::unordered_map<std::string, int> stateNames;
+            for (auto [k, state] : phaseContext["States"]) {
+                stateNames.emplace(state["Name"].to<std::string>(), i);
+                i++;
+            }
+            auto numStates = stateNames.size();
+            stateConfig.continueCurrentState =
+                phaseContext["Continue"].maybe<bool>().value_or(false);
+            if (stateConfig.continueCurrentState) {
+                // Check that this isn't phase 0
+                auto myPhaseNumber = phaseContext.getPhaseNumber();
+                if (myPhaseNumber == 0) {
+                    BOOST_THROW_EXCEPTION(InvalidConfigurationException(
+                        "CrudActor has continue set for phase 0. Nothing to continue from."));
+                }
+                // TODO: check that the previous state has <= number of states as this one
+                // Need to access the previous phase config.
+
+            } else if (!phaseContext["InitialStates"]) {
+                BOOST_THROW_EXCEPTION(InvalidConfigurationException(
+                    "CrudActor has States, but has not specified InitialStates"));
+            }
+
+            stateConfig.states.reserve(numStates);
+            for (auto [k, state] : phaseContext["States"]) {
+                stateConfig.states.emplace_back(addState(state, stateNames));
+            }
+            stateConfig.initialStateWeights = std::vector<double>(numStates, 0);
+            for (auto [k, state] : phaseContext["InitialStates"]) {
+                stateConfig.initialStateWeights[stateNames[state["State"].to<std::string>()]] +=
+                    state["Weight"].to<double>();
+            }
+        } else {  // Throw a useful error}
+            BOOST_THROW_EXCEPTION(
+                InvalidConfigurationException("CrudActor has neither Operations nor States "
+                                              "specified. Exactly one must be defined."));
+        }
     }
 };
 
 void CrudActor::run() {
+    int currentState = 0;
+    int nextState = 0;
     for (auto&& config : _loop) {
         auto session = _client->start_session();
+        // TODO: If running without states, define a single state with the operations
+        if (!config.isNop() && !config->stateConfig.states.empty() &&
+            !config->stateConfig.continueCurrentState) {
+            // pick the initial state for the phase
+            auto initial_distribution =
+                boost::random::discrete_distribution(config->stateConfig.initialStateWeights);
+            nextState = initial_distribution(_rng);
+            BOOST_LOG_TRIVIAL(debug) << "Actor " << id() << " picking initial state " << nextState;
+        }
         for (const auto&& _ : config) {
             auto metricsContext = config->metrics.start();
-
-            for (auto&& op : config->operations) {
-                op->run(session);
+            if (!config->stateConfig.states.empty()) {
```

Comment:
This seems like a big and possibly difficult to maintain change. 

At the very least *advanceFSM* would help hide it away (and maybe most of the FSM code ends up in stateConfig).